File: ts/indicators/donald-lambert/commodity-channel-index/commodity-channel-index.ts
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorType } from '../../core/indicator-type';
import { LineIndicator } from '../../core/line-indicator';
import { OutputType } from '../../core/outputs/output-type';
import { CommodityChannelIndexOutput } from './commodity-channel-index-output';
import { CommodityChannelIndexParams, DefaultInverseScalingFactor } from './commodity-channel-index-params';

/** Function to calculate mnemonic of a __CommodityChannelIndex__ indicator. */
export const commodityChannelIndexMnemonic = (params: CommodityChannelIndexParams): string => {
  const cm = componentTripleMnemonic(
    params.barComponent,
    params.quoteComponent,
    params.tradeComponent,
  );

  return `cci(${params.length}${cm})`;
};

/**
 * CommodityChannelIndex is Donald Lambert's Commodity Channel Index (CCI).
 *
 * CCI measures the deviation of the price from its statistical mean. High values
 * indicate that prices are unusually high compared to average, and low values
 * indicate that prices are unusually low.
 *
 *   CCI = (typicalPrice - SMA) / (scalingFactor * meanDeviation)
 *
 * where scalingFactor defaults to 0.015 so that approximately 70-80% of CCI values
 * fall between -100 and +100.
 *
 * Reference:
 *
 * Lambert, Donald (1980). "Commodity Channel Index: Tools for Trading Cyclic Trends".
 * Commodities (now Futures) magazine.
 */
export class CommodityChannelIndex extends LineIndicator {
  private readonly length: number;
  private readonly scalingFactor: number;
  private readonly window: number[];
  private windowCount: number;
  private windowSum: number;
  private value: number;

  /**
   * Constructs an instance given the parameters.
   */
  public constructor(params: CommodityChannelIndexParams) {
    super();

    const length = Math.floor(params.length);

    if (length < 2) {
      throw new Error('length should be greater than 1');
    }

    const inverseFactor = params.inverseScalingFactor ?? DefaultInverseScalingFactor;

    this.length = length;
    this.scalingFactor = length / inverseFactor;
    this.window = new Array<number>(length).fill(0);
    this.windowCount = 0;
    this.windowSum = 0;
    this.value = Number.NaN;
    this.primed = false;

    this.mnemonic = commodityChannelIndexMnemonic(params);
    this.description = 'Commodity Channel Index ' + this.mnemonic;
    this.barComponent = params.barComponent;
    this.quoteComponent = params.quoteComponent;
    this.tradeComponent = params.tradeComponent;
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return {
      type: IndicatorType.CommodityChannelIndex,
      mnemonic: this.mnemonic,
      description: this.description,
      outputs: [{
        kind: CommodityChannelIndexOutput.CommodityChannelIndexValue,
        type: OutputType.Scalar,
        mnemonic: this.mnemonic,
        description: this.description,
      }],
    };
  }

  /** Updates the value of the indicator given the next sample. */
  public update(sample: number): number {
    if (Number.isNaN(sample)) {
      return sample;
    }

    const lastIndex = this.length - 1;

    if (this.primed) {
      this.windowSum += sample - this.window[0];

      for (let i = 0; i < lastIndex; i++) {
        this.window[i] = this.window[i + 1];
      }

      this.window[lastIndex] = sample;

      const average = this.windowSum / this.length;

      let temp = 0;
      for (let i = 0; i < this.length; i++) {
        temp += Math.abs(this.window[i] - average);
      }

      if (Math.abs(temp) < Number.EPSILON) {
        this.value = 0;
      } else {
        this.value = this.scalingFactor * (sample - average) / temp;
      }
    } else {
      this.windowSum += sample;
      this.window[this.windowCount] = sample;
      this.windowCount++;

      if (this.windowCount === this.length) {
        this.primed = true;

        const average = this.windowSum / this.length;

        let temp = 0;
        for (let i = 0; i < this.length; i++) {
          temp += Math.abs(this.window[i] - average);
        }

        if (Math.abs(temp) < Number.EPSILON) {
          this.value = 0;
        } else {
          this.value = this.scalingFactor * (sample - average) / temp;
        }
      }
    }

    return this.value;
  }
}

File: go/indicators/donaldlambert/commoditychannelindex/commoditychannelindex.go
package commoditychannelindex

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
)

// CommodityChannelIndex is Donald Lambert's Commodity Channel Index (CCI).
//
// CCI measures the deviation of the price from its statistical mean. High values
// indicate that prices are unusually high compared to average, and low values
// indicate that prices are unusually low.
//
//	CCI = (typicalPrice - SMA) / (scalingFactor * meanDeviation)
//
// where scalingFactor defaults to 0.015 so that approximately 70-80% of CCI values
// fall between -100 and +100.
//
// Reference:
//
// Lambert, Donald (1980). "Commodity Channel Index: Tools for Trading Cyclic Trends".
// Commodities (now Futures) magazine.
type CommodityChannelIndex struct {
	mu sync.RWMutex
	core.LineIndicator
	length        int
	scalingFactor float64
	window        []float64
	windowCount   int
	windowSum     float64
	value         float64
	primed        bool
}

// NewCommodityChannelIndex returns an instance of the indicator created using supplied parameters.
func NewCommodityChannelIndex(p *CommodityChannelIndexParams) (*CommodityChannelIndex, error) {
	const (
		invalid   = "invalid commodity channel index parameters"
		fmts      = "%s: %s"
		fmtw      = "%s: %w"
		minLength = 2
	)

	if p.Length < minLength {
		return nil, fmt.Errorf(fmts, invalid, "length should be greater than 1")
	}

	inverseFactor := p.InverseScalingFactor
	if inverseFactor == 0 {
		inverseFactor = DefaultInverseScalingFactor
	}

	bc := p.BarComponent
	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	qc := p.QuoteComponent
	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	tc := p.TradeComponent
	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	mnemonic := fmt.Sprintf("cci(%d%s)", p.Length,
		core.ComponentTripleMnemonic(bc, qc, tc))
	desc := "Commodity Channel Index " + mnemonic

	cci := &CommodityChannelIndex{
		length:        p.Length,
		scalingFactor: float64(p.Length) / inverseFactor,
		window:        make([]float64, p.Length),
		value:         math.NaN(),
	}

	cci.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, cci.Update)

	return cci, nil
}

// IsPrimed indicates whether the indicator is primed.
func (s *CommodityChannelIndex) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes the output data of the indicator.
func (s *CommodityChannelIndex) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.CommodityChannelIndex,
		s.LineIndicator.Mnemonic,
		s.LineIndicator.Description,
		[]core.OutputText{
			{Mnemonic: s.LineIndicator.Mnemonic, Description: s.LineIndicator.Description},
		},
	)
}

// Update updates the value of the indicator given the next sample.
func (s *CommodityChannelIndex) Update(sample float64) float64 {
	if math.IsNaN(sample) {
		return sample
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	lastIndex := s.length - 1

	if s.primed {
		s.windowSum += sample - s.window[0]
		copy(s.window, s.window[1:])
		s.window[lastIndex] = sample

		average := s.windowSum / float64(s.length)

		var temp float64
		for i := 0; i < s.length; i++ {
			temp += math.Abs(s.window[i] - average)
		}

		if math.Abs(temp) < math.SmallestNonzeroFloat64 {
			s.value = 0
		} else {
			s.value = s.scalingFactor * (sample - average) / temp
		}
	} else {
		s.windowSum += sample
		s.window[s.windowCount] = sample
		s.windowCount++

		if s.windowCount == s.length {
			s.primed = true

			average := s.windowSum / float64(s.length)

			var temp float64
			for i := 0; i < s.length; i++ {
				temp += math.Abs(s.window[i] - average)
			}

			if math.Abs(temp) < math.SmallestNonzeroFloat64 {
				s.value = 0
			} else {
				s.value = s.scalingFactor * (sample - average) / temp
			}
		}
	}

	return s.value
}
